homeassistant/components/rointe/coordinator.py

```diff
@@ -0,0 +1,108 @@
+"""Provides the Rointe DataUpdateCoordinator."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import timedelta
+import logging
+from typing import Any
+
+from rointesdk.device import RointeDevice
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import device_registry
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, PLATFORMS, ROINTE_API_REFRESH_SECONDS
+from .device_manager import RointeDeviceManager
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class RointeDataUpdateCoordinator(DataUpdateCoordinator[dict[str, RointeDevice]]):
+    """Rointe data coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, device_manager: RointeDeviceManager
+    ) -> None:
+        """Initialize Rointe data updater."""
+        self.device_manager = device_manager
+        self.unregistered_keys: dict[str, dict[str, RointeDevice]] = {}
+        self.cleanup_callbacks: list[Callable[[], None]] = []
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=ROINTE_API_REFRESH_SECONDS),
+        )
+
+        for platform in PLATFORMS:
+            self.unregistered_keys[platform] = {}
+
+    async def _async_update_data(self) -> dict[str, RointeDevice]:
+        """Fetch data from API."""
+
+        devices = await self.device_manager.update()
+
+        for device_id, device in devices.items():
+            for platform in PLATFORMS:
+                if device_id not in self.unregistered_keys[platform]:
+                    self.unregistered_keys[platform][device_id] = device
```

Comment:
Nested loops indicate a good spot for a dict comprehension. I can't tell your logic, so I can't give you a specific example.

Reply:
Refactored this block and split the call to `device_update_info` outside the loops. The purpose of that call is to update the registry in case a device's firmware is updated (Otherwise it will keep showing as "update available")